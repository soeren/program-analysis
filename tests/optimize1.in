// Check for program optimization
// - Check that sequence of constant folding plus dead code
// elimination removes fainted variables and reduces code to a minimum

program

int x;
int y;
int z;

int identity(int b) {
    return b;
}

int givethree(int a) {
    int t;
    int r;
    t := 3;
    r := a*t;
    return t;
}

int diff(int a, int b){
    int s;
    int y;
    y := a - b;
    s := y;
    return s;
}

x := diff(4,2);
y := 3;

if(y > x){
    x := x*identity(y);
} else {
    x := 0;
}

while(x > 0){
    y := y * givethree(0);
}

x := y - x;

write(x);

end
