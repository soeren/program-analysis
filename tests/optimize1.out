program
int x;
int y;
y := 3; // 1
while(true) // 2
{
     y := y * 3; // 3
 }
x := y - 6; // 4
write(x); // 5
end
